### ISSUE TITLE
Adding/removing/renaming a class always forces dependent source file to recompile

### ISSUE THREAD

**Issue by gkossakowski** [sbt/sbt#2319]
`A.scala`

``` scala
// class Q

object A { val x = 3 }
```

`B.scala`

``` scala
object B { val x = A.x }
```
1. `compile`
2. Uncomment line in `A.scala`.
3. `compile`. Both sources are recompiled because `Q` introduces a change to the hash for `<init>` (in the compilation unit `A.scala`).

The underlying reason is that name hashing is tracking hashes for names at source file (compilation unit) level. The hash for `<init>` includes hashes for both `A.<init>` and `Q.<init>`. Therefore name hashing works well only when a member that is modified doesn't introduce any nested members. Therefore, it works for changes to `val` and `def` but not to `class` and `object`.

**Comment by gkossakowski** [comment]
This issue is migration of gkossakowski/sbt#8.

**Comment by adriaanm** [comment]
[for <init>] in the compilation unit A.scala.

**Comment by jvican** [comment]
This can be closed since it's fixed in sbt 1.0.0.
